### ISSUE TITLE
[求助] 如何让模型输出对下一个token的概率分布预测？

### ISSUE THREAD

**Issue by LeoLoudWarden** [QwenLM/Qwen1.5#232]
尝试处理predict_next任务，思路是设置 return_dict_in_generate=True 以及 output_scores=True
运行以下代码，并不能得到预期中的效果。
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
model_path = "E:\\nlp_models\\Qwen1.5-7B-Chat-GPTQ-Int4"
# model_path = "E:\\nlp_models\\Qwen1.5-7B-Chat"
device = "cuda"

tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
    model_path,
    torch_dtype='auto',
    device_map=device,
).eval()

prompt = "Give me a short introduction to large language model."
messages = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": prompt}
]
text = tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=True
)
model_inputs = tokenizer([text], return_tensors="pt").to(device)

output = model.generate(
    model_inputs.input_ids,
    max_new_tokens=1,
    return_dict_in_generate=True,
    output_scores=True,
)

print(output['scores'])
```
输出为:
```
(tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]], device='cuda:0'),)
```
然而，对于ChatGLM3-6B采用相同的参数model.generate，是可行的。问题出在哪里？

**Comment by jklj077** [comment]
The represenation of the tensor only shows the starting (`!`, `"`, and `#`) and the ending tokens (invalid token indices, padded for efficency) in the vocabulary, which are unlikely to be meaningful for that position. Apply softmax to it and there should be meaningful values in the result.

You may also need to tune the hyper-parameters in `generation_config.json` such that the distribution is less concentrated and refrain from using the quantized or chat models.

**Comment by zoeChen119** [comment]
我也是

**Comment by jklj077** [comment]
As per my last reply, here is the script and the outputs from Qwen1.5-0.5B
```python
>>> tokens = tokenizer(tokenizer.apply_chat_template([{"role": "user", "content": "你好"}], tokenize=False, add_generation_prompt=True), return_tensors="pt")
>>> tokens
{'input_ids': tensor([[151644,   8948,    198,   2610,    525,    264,  10950,  17847, 151645,
            198, 151644,    872,    198, 108386, 151645,    198, 151644,  77091,
            198]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

>>> output = model.generate(**tokens, return_dict_in_generate=True, output_scores=True, max_new_tokens=10)
>>> output.sequences
tensor([[151644,   8948,    198,   2610,    525,    264,  10950,  17847, 151645,
            198, 151644,    872,    198, 108386, 151645,    198, 151644,  77091,
            198, 111308,   6313, 104139, 110913,  99663, 101214, 101037,  11319,
         151645]])
>>> output.scores
(tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
 tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]))

>>> [{i: p for i, p in enumerate(prob[0]) if p > 0} for prob in [torch.softmax(score, dim=1).tolist() for score in output.scores]]
[{108386: 0.6088112592697144, 111308: 0.3911888003349304},
 {6313: 1.0},
 {87026: 0.10710759460926056,
  104139: 0.7604665756225586,
  112169: 0.13242581486701965},
 {109944: 0.45277032256126404,
  110913: 0.33639588952064514,
  111728: 0.2108338326215744},
 {99663: 0.6459924578666687,
  100364: 0.25878724455833435,
  102804: 0.09522030502557755},
 {101214: 1.0},
 {101037: 1.0},
 {11319: 1.0},
 {151645: 1.0}]
```

**Comment by lucy9527** [comment]
> As per my last reply, here is the script and the outputs from Qwen1.5-0.5B-Chat
> 
> ```python
> >>> tokens = tokenizer(tokenizer.apply_chat_template([{"role": "user", "content": "你好"}], tokenize=False, add_generation_prompt=True), return_tensors="pt")
> >>> tokens
> {'input_ids': tensor([[151644,   8948,    198,   2610,    525,    264,  10950,  17847, 151645,
>             198, 151644,    872,    198, 108386, 151645,    198, 151644,  77091,
>             198]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}
> 
> >>> output = model.generate(**tokens, return_dict_in_generate=True, output_scores=True, max_new_tokens=10)
> >>> output.sequences
> tensor([[151644,   8948,    198,   2610,    525,    264,  10950,  17847, 151645,
>             198, 151644,    872,    198, 108386, 151645,    198, 151644,  77091,
>             198, 111308,   6313, 104139, 110913,  99663, 101214, 101037,  11319,
>          151645]])
> >>> output.scores
> (tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]))
> 
> >>> [{i: p for i, p in enumerate(prob[0]) if p > 0} for prob in [torch.softmax(score, dim=1).tolist() for score in output.scores]]
> [{108386: 0.6088112592697144, 111308: 0.3911888003349304},
>  {6313: 1.0},
>  {87026: 0.10710759460926056,
>   104139: 0.7604665756225586,
>   112169: 0.13242581486701965},
>  {109944: 0.45277032256126404,
>   110913: 0.33639588952064514,
>   111728: 0.2108338326215744},
>  {99663: 0.6459924578666687,
>   100364: 0.25878724455833435,
>   102804: 0.09522030502557755},
>  {101214: 1.0},
>  {101037: 1.0},
>  {11319: 1.0},
>  {151645: 1.0}]
> ```

Nice brain

**Comment by lucy9527** [comment]
> As per my last reply, here is the script and the outputs from Qwen1.5-0.5B-Chat
> 
> ```python
> >>> tokens = tokenizer(tokenizer.apply_chat_template([{"role": "user", "content": "你好"}], tokenize=False, add_generation_prompt=True), return_tensors="pt")
> >>> tokens
> {'input_ids': tensor([[151644,   8948,    198,   2610,    525,    264,  10950,  17847, 151645,
>             198, 151644,    872,    198, 108386, 151645,    198, 151644,  77091,
>             198]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}
> 
> >>> output = model.generate(**tokens, return_dict_in_generate=True, output_scores=True, max_new_tokens=10)
> >>> output.sequences
> tensor([[151644,   8948,    198,   2610,    525,    264,  10950,  17847, 151645,
>             198, 151644,    872,    198, 108386, 151645,    198, 151644,  77091,
>             198, 111308,   6313, 104139, 110913,  99663, 101214, 101037,  11319,
>          151645]])
> >>> output.scores
> (tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]),
>  tensor([[-inf, -inf, -inf,  ..., -inf, -inf, -inf]]))
> 
> >>> [{i: p for i, p in enumerate(prob[0]) if p > 0} for prob in [torch.softmax(score, dim=1).tolist() for score in output.scores]]
> [{108386: 0.6088112592697144, 111308: 0.3911888003349304},
>  {6313: 1.0},
>  {87026: 0.10710759460926056,
>   104139: 0.7604665756225586,
>   112169: 0.13242581486701965},
>  {109944: 0.45277032256126404,
>   110913: 0.33639588952064514,
>   111728: 0.2108338326215744},
>  {99663: 0.6459924578666687,
>   100364: 0.25878724455833435,
>   102804: 0.09522030502557755},
>  {101214: 1.0},
>  {101037: 1.0},
>  {11319: 1.0},
>  {151645: 1.0}]
> ```

nice coding, but it‘s useful when do_sample==false. When there is sampling randomness, should show as follow:
```
{1: 0.08751006424427032, 2073: 0.08751006424427032, 18493: 0.04400281980633736, 99513: 0.08751006424427032, 99692: 0.03647962212562561, 99880: 0.041336823254823685, 102332: 0.04684076085686684, 104576: 0.49320343136787415, 104596: 0.041336823254823685, 110834: 0.034269433468580246}
{104576: 1.0}
{99530: 1.0}
{1: 0.5504235029220581, 3837: 0.4495764672756195}
{3837: 0.4441203474998474, 6567: 0.01722041331231594, 9370: 0.01722041331231594, 20412: 0.0874524712562561, 26939: 0.17392002046108246, 54851: 0.08215400576591492, 99495: 0.015196962282061577, 99639: 0.025055555626749992, 101909: 0.10548783093690872, 104053: 0.03217197209596634}
{3837: 0.5926666259765625, 104060: 0.4073334336280823}
{3837: 1.0}
{99513: 1.0}
{56568: 0.5856949090957642, 87026: 0.045163944363594055, 99466: 0.16780440509319305, 103929: 0.17862685024738312, 107740: 0.022709855809807777}
{99424: 0.9241418242454529, 114169: 0.07585817575454712}
{29524: 0.5858044624328613, 49111: 0.020045125856995583, 102216: 0.22940431535243988, 102694: 0.033048827201128006, 104576: 0.08591372519731522, 106837: 0.02573845162987709, 114219: 0.020045125856995583}
```
